File: java-workshop/part-3-mocks-clean/src/test/java/service/OrderCancellationServiceTest.java
package service;

import domain.CancellationResult;
import domain.Order;
import domain.OrderStatus;
import domain.StockReservation;
import helper.OrderBuilder;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;
import repository.OrderRepository;

import java.util.List;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.anyString;
import static org.mockito.Mockito.never;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.verifyNoInteractions;
import static org.mockito.Mockito.when;

@ExtendWith(MockitoExtension.class)
class OrderCancellationServiceTest {

    @Mock
    private OrderRepository orderRepository;

    @Mock
    private PaymentService paymentService;

    @Mock
    private StockService stockService;

    @Mock
    private ShippingService shippingService;

    @InjectMocks
    private OrderCancellationService service;

    private Order confirmedOrder;

    @BeforeEach
    void setUp() {
        confirmedOrder = OrderBuilder.confirmed()
            .withId("ORDER-001")
            .withTransaction("TXN-123")
            .withTracking("TRACK-456")
            .build();
    }

    @Test
    void cancelOrder_succeeds_whenAllStepsPass() {
        givenOrderExists(confirmedOrder);
        givenRefundSucceeds();
        givenShipmentCancelled();

        CancellationResult result = service.cancelOrder(confirmedOrder.id());

        assertSuccess(result);
        verify(paymentService).refundPayment("TXN-123");
        verify(shippingService).cancelShipment("TRACK-456");
        verifyStockReleasedFor(confirmedOrder);
        verify(orderRepository).updateStatus("ORDER-001", OrderStatus.CANCELLED);
    }

    @Test
    void cancelOrder_fails_whenRefundFails() {
        givenOrderExists(confirmedOrder);
        givenRefundFails();

        CancellationResult result = service.cancelOrder("ORDER-001");

        assertFailure(result, "Refund failed");
        verify(paymentService).refundPayment("TXN-123");
        verify(orderRepository, never()).updateStatus(anyString(), any());
        verifyNoInteractions(shippingService, stockService);
    }

    @Test
    void cancelOrder_skipsShipmentCancellation_whenOrderHasNoTrackingNumber() {
        Order orderWithoutShipment = OrderBuilder.confirmed()
            .withId("ORDER-002")
            .withTransaction("TXN-456")
            .withoutTracking()
            .build();

        givenOrderExists(orderWithoutShipment);
        givenRefundSucceeds();

        CancellationResult result = service.cancelOrder(orderWithoutShipment.id());

        assertSuccess(result);
        verify(paymentService).refundPayment("TXN-456");
        verifyNoInteractions(shippingService);
        verifyStockReleasedFor(orderWithoutShipment);
        verify(orderRepository).updateStatus("ORDER-002", OrderStatus.CANCELLED);
    }

    private void givenOrderExists(Order order) {
        when(orderRepository.findById(order.id())).thenReturn(order);
    }

    private void givenRefundSucceeds() {
        when(paymentService.refundPayment(anyString())).thenReturn(true);
    }

    private void givenRefundFails() {
        when(paymentService.refundPayment(anyString())).thenReturn(false);
    }

    private void givenShipmentCancelled() {
        when(shippingService.cancelShipment(anyString())).thenReturn(true);
    }

    private void verifyStockReleasedFor(Order order) {
        List<StockReservation> expectedReservations = order.items().stream()
            .map(item -> new StockReservation(item.productId(), item.quantity(), true))
            .toList();

        verify(stockService).releaseStock(expectedReservations);
    }

    private void assertSuccess(CancellationResult result) {
        assertEquals(new CancellationResult.Success(), result);
    }

    private void assertFailure(CancellationResult result, String reason) {
        assertEquals(new CancellationResult.Failure(reason), result);
    }
}

File: csharp-workshop/Part3.MocksClean.Tests/Service/OrderCancellationServiceTest.cs
using NSubstitute;
using Part3.MocksClean.Domain;
using Part3.MocksClean.Repository;
using Part3.MocksClean.Service;
using Part3.MocksClean.Tests.Helper;
using Xunit;

namespace Part3.MocksClean.Tests.Service;

public class OrderCancellationServiceTest
{
    private readonly IOrderRepository _orderRepository = Substitute.For<IOrderRepository>();
    private readonly IPaymentService _paymentService = Substitute.For<IPaymentService>();
    private readonly IStockService _stockService = Substitute.For<IStockService>();
    private readonly IShippingService _shippingService = Substitute.For<IShippingService>();
    private readonly OrderCancellationService _service;
    private readonly Order _confirmedOrder;

    public OrderCancellationServiceTest()
    {
        _service = new OrderCancellationService(_orderRepository, _paymentService, _stockService, _shippingService);
        _confirmedOrder = OrderBuilder.Confirmed()
            .WithId("ORDER-001")
            .WithTransaction("TXN-123")
            .WithTracking("TRACK-456")
            .Build();
    }

    [Fact]
    public void P3_CancelOrder_Succeeds_WhenAllStepsPass()
    {
        GivenOrderExists(_confirmedOrder);
        GivenRefundSucceeds();
        GivenShipmentCancelled();

        CancellationResult result = _service.CancelOrder(_confirmedOrder.Id);

        AssertSuccess(result);
        _paymentService.Received(1).RefundPayment("TXN-123");
        _shippingService.Received(1).CancelShipment("TRACK-456");
        VerifyStockReleasedFor(_confirmedOrder);
        _orderRepository.Received(1).UpdateStatus("ORDER-001", OrderStatus.Cancelled);
    }

    [Fact]
    public void P3_CancelOrder_Fails_WhenRefundFails()
    {
        GivenOrderExists(_confirmedOrder);
        GivenRefundFails();

        CancellationResult result = _service.CancelOrder("ORDER-001");

        AssertFailure(result, "Refund failed");
        _paymentService.Received(1).RefundPayment("TXN-123");
        _orderRepository.DidNotReceive().UpdateStatus(Arg.Any<string>(), Arg.Any<OrderStatus>());
        _shippingService.DidNotReceiveWithAnyArgs();
        _stockService.DidNotReceiveWithAnyArgs();
    }

    [Fact]
    public void P3_CancelOrder_SkipsShipmentCancellation_WhenOrderHasNoTrackingNumber()
    {
        var orderWithoutShipment = OrderBuilder.Confirmed()
            .WithId("ORDER-002")
            .WithTransaction("TXN-456")
            .WithoutTracking()
            .Build();

        GivenOrderExists(orderWithoutShipment);
        GivenRefundSucceeds();

        CancellationResult result = _service.CancelOrder(orderWithoutShipment.Id);

        AssertSuccess(result);
        _paymentService.Received(1).RefundPayment("TXN-456");
        _shippingService.DidNotReceiveWithAnyArgs();
        VerifyStockReleasedFor(orderWithoutShipment);
        _orderRepository.Received(1).UpdateStatus("ORDER-002", OrderStatus.Cancelled);
    }

    [Fact]
    public void P3_CancelOrder_Fails_WhenOrderDoesNotExist()
    {
        _orderRepository.FindById("ORDER-UNKNOWN").Returns((Order?)null);

        CancellationResult result = _service.CancelOrder("ORDER-UNKNOWN");

        AssertFailure(result, "Order not found");
        _paymentService.DidNotReceiveWithAnyArgs();
        _stockService.DidNotReceiveWithAnyArgs();
        _shippingService.DidNotReceiveWithAnyArgs();
    }

    private void GivenOrderExists(Order order)
    {
        _orderRepository.FindById(order.Id).Returns(order);
    }

    private void GivenRefundSucceeds()
    {
        _paymentService.RefundPayment(Arg.Any<string>()).Returns(true);
    }

    private void GivenRefundFails()
    {
        _paymentService.RefundPayment(Arg.Any<string>()).Returns(false);
    }

    private void GivenShipmentCancelled()
    {
        _shippingService.CancelShipment(Arg.Any<string>()).Returns(true);
    }

    private void VerifyStockReleasedFor(Order order)
    {
        var expectedReservations = order.Items
            .Select(item => new StockReservation(item.ProductId, item.Quantity, true))
            .ToList();
        _stockService.Received(1).ReleaseStock(expectedReservations);
    }

    private static void AssertSuccess(CancellationResult result)
    {
        Assert.Equal(new CancellationResult.Success(), result);
    }

    private static void AssertFailure(CancellationResult result, string reason)
    {
        Assert.Equal(new CancellationResult.Failure(reason), result);
    }
}
